src/test/ui/parser-recovery-3.stderr

```diff
@@ -0,0 +1,41 @@
+error: incorrect close delimiter: `}`
+  --> $DIR/parser-recovery-3.rs:19:5
+   |
+LL |     fn banana(&mut self) {
+   |                          - close delimiter possibly meant for this
+LL |         fn peach(this: &Self, foo: usize {
+   |                 - un-closed delimiter
+...
+LL |     }
+   |     ^ incorrect close delimiter
+
+error: expected one of `!`, `(`, `)`, `+`, `,`, `::`, or `<`, found `{`
```

Comment:
it's kind of surprising to me that this advice comes here, in a second error after the "main" error above... why is that?

Reply:
The lexer first performs the basic matching braces checks and emits the appropriate error. _Then_ the parser takes the tokenstream and tries to parse things. So the first error is about the lexer finding unbalanced delimiters, while the second one is a parse error not finding an expected close delimiter.

There're two options: keeping the un-emitted DiagnosticBuilders from the lexer and emitting them once we are parsing and find an appropriate place for the close delimiter suggestion, or completely stop performing balanced delimiter checks in the lexer and rely entirely on the parser for parse errors. Both felt like too much work for little win. I'd prefer to iterate this PR until we have the recovery nailed down before revisiting the above possibilities.